### PR TITLE
fix (re-implement) gpu binding in mpibind.c

### PR DESCRIPTION
**Pull request by ryanday36**
This fixes the parsing of hwloc objects so that it actually discovers the correct number of gpus and then uses the algorithm from Edgar Leon Borja's mpibind8 to assign gpus to tasks.

### CODE REVIEW

File: mpibind.c [hunk 1]
```diff
@@ -515,24 +509,120 @@ static char *get_gomp_str (hwloc_cpuset_t cpuset)
...
+static char *get_gpustring( int32_t gpus, int32_t numas, uint32_t *numagroup )
 {
     char *str = NULL;
-    int32_t i, rc;
+    hwloc_obj_t obj;
+    int32_t mapped_numa=0, mapped_np_in_numa=0, mapped_id_in_numa=0;
+    uint32_t (*gpus_in_numa)[gpus] = malloc(sizeof(uint32_t[numas][gpus]));
+    uint32_t (*gpus_in_group)[gpus] = malloc(sizeof(uint32_t[numas][gpus]));
+    uint32_t *gpus_per_numa = calloc (numas, sizeof (uint32_t));
+    uint32_t *gpus_per_group = calloc (numas, sizeof (uint32_t));
```
Comment (grondo) [inline mpibind.c:524]: Maybe a leak here, are these allocations freed before exiting the function? Sorry if I just missed it.

**Review by grondo — COMMENTED** [review]
LGTM! Man this looks like a huge amount of work.

I had just one question about possible memleak in `get_gpustring`, though a leak here is low impact since the plugin only runs once.

File: mpibind.c (hunk 1 above)
Comment (ryanday36) [reply mpibind.c:524]: You didn't miss anything. I've just spent all of my time in scripting languages for too long. I'll free those and submit a new PR.

Comment (ryanday36) [reply mpibind.c:524]: 'frees' committed. Guess I don't need to submit a new PR.

**Comment by grondo** [comment]
Cool. You can squash your commits together if you want, or merge as is. Just force a push to your branch and the PR will be updated

**Comment by ryanday36** [comment]
Alright. I think everything is pushed.

**Comment by grondo** [comment]
@ryanday36, ok. The commit to fix the memleak is still separate from the main commit here. History might be cleaner if you squash that one as a fixup into your first commit, i.e.
```console
$ git rebase -i HEAD~2  # interactive rebase last two commits
````

Then mark the last commit as a fixup to the first.
```
pick 926f495 fix (re-implement) gpu binding in mpibind.c
f 4c0ce28 free up allocated memory
```

Then force push the result

```console
$ git push -f origin master
```

**Comment by ryanday36** [comment]
Ah. That's handy. I think it's all there now.

**Comment by grondo** [comment]
Ok, merged! Feel free to push a new tag directly .

**Comment by ryanday36** [comment]
should I call it 0.38 and update NEWS and META accordingly?

**Comment by grondo** [comment]
Yeah -- I'm actually not sure the META is used anymore, but in case it is best to update it :stuck_out_tongue_winking_eye:

**Pull request merged**